github/resource_github_actions_secret.go

```diff
@@ -0,0 +1,166 @@
+package github
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"github.com/google/go-github/v29/github"
+	"github.com/hashicorp/terraform/helper/schema"
+	"golang.org/x/crypto/nacl/box"
+	"log"
+)
+
+func resourceGithubActionsSecret() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGithubActionsSecretCreateOrUpdate,
+		Read:   resourceGithubActionsSecretRead,
+		Update: resourceGithubActionsSecretCreateOrUpdate,
+		Delete: resourceGithubActionsSecretDelete,
+
+		Schema: map[string]*schema.Schema{
+			"repository": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"secret_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"plaintext_value": {
+				Type:      schema.TypeString,
+				Required:  true,
+				Sensitive: true,
+			},
+			"created_at": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"updated_at": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceGithubActionsSecretCreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	err := checkOrganization(meta)
+	if err != nil {
+		return err
+	}
+
+	client := meta.(*Organization).client
+	owner := meta.(*Organization).name
+	ctx := context.Background()
+
+	repo := d.Get("repository").(string)
+	secretName := d.Get("secret_name").(string)
+	plaintextValue := d.Get("plaintext_value").(string)
+
+	keyId, publicKey, err := getPublicKeyDetails(owner, repo, meta)
+	if err != nil {
+		return err
+	}
+
+	encryptedText, err := encryptPlaintext(plaintextValue, publicKey)
+	if err != nil {
+		return err
+	}
+
+	// Create an EncryptedSecret and encrypt the plaintext value into it
+	eSecret := &github.EncryptedSecret{
+		Name:           secretName,
+		KeyID:          keyId,
+		EncryptedValue: base64.StdEncoding.EncodeToString(encryptedText),
+	}
+
+	_, err = client.Actions.CreateOrUpdateSecret(ctx, owner, repo, eSecret)
+	if err != nil {
+		return err
+	}
+
+	return resourceGithubActionsSecretRead(d, meta)
+}
+
+func resourceGithubActionsSecretRead(d *schema.ResourceData, meta interface{}) error {
+	err := checkOrganization(meta)
+	if err != nil {
+		return err
+	}
+
+	client := meta.(*Organization).client
+	owner := meta.(*Organization).name
+	ctx := context.Background()
+
+	repo := d.Get("repository").(string)
+	secretName := d.Get("secret_name").(string)
+
+	secret, _, err := client.Actions.GetSecret(ctx, owner, repo, secretName)
+	if err != nil || secret.Name != secretName {
+		d.SetId("")
+		return err
+	}
+
+	d.SetId(buildTwoPartID(&repo, &secretName))
+	d.Set("plaintext_value", d.Get("plaintext_value"))
```

Comment:
Not 100% sure but I think is not required, as it already should exist in state.

Reply:
I am also not sure on this point. I left it there such that it was obvious (and definite) it would be set. I don't see any issue with leaving this line in?